proxy/src/transport/tls/conditional_accept.rs

```diff
@@ -0,0 +1,248 @@
+#![allow(dead_code)] // TODO: Use this.
+
+use super::{Identity, untrusted};
+
+#[derive(Debug, Eq, PartialEq)]
+pub enum Match {
+    Incomplete,
+    Matched,
+    NotMatched,
+}
+
+/// Determintes whether the given `input` looks like the start of a TLS
+/// connection that the proxy should terminate.
+///
+/// The determination is made based on whether the input looks like (the start
+/// of) a valid ClientHello that a reasonable TLS client might send, and the
+/// SNI matches the given identity.
+///
+/// XXX: Once the TLS record header is matched, the determination won't be
+/// made until the entire TLS record including the entire ClientHello handshake
+/// message is available. TODO: Reject non-matching inputs earlier.
+///
+/// This assumes that the ClientHello is small and is sent in a single TLS
+/// record, which is what all reasonable implementations do. (If they were not
+/// to, they wouldn't interoperate with picky servers.)
+pub fn match_client_hello(input: &[u8], identity: &Identity) -> Match {
+    let r = untrusted::Input::from(input).read_all(untrusted::EndOfInput, |input| {
+        let r = extract_sni(input);
+        input.skip_to_end(); // Ignore anything after what we parsed.
+        r
+    });
+    match r {
+        Ok(Some(sni)) => {
+            let matches = if let Ok(sni) = Identity::from_sni_hostname(sni.as_slice_less_safe()) {
```

Comment:
Fair. It is explained in the `untrusted` documentation at https://briansmith.org/rustdoc/untrusted/: "untrusted.rs goes beyond Rust's normal safety guarantees by also guaranteeing that parsing will be panic-free, as long as untrusted::Input::as_slice_less_safe() is not used." Basically when using `untrusted` we try to expose the input only as `untrusted::Input` and `untrusted::Reader` to ensure there are no bounds checking panics. When we can't do that, e.g. because we need to interact with an API that doesn't take `untrusted::Input` as a parameter, then we have to give up the panic-safety, and that's what this name indicates.